lib/location/method_wifi.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <zephyr.h>
+#include <assert.h>
+
+#include <logging/log.h>
+
+#include <net/net_if.h>
+#include <net/net_event.h>
+#include <net/wifi_mgmt.h>
+
+#include <modem/location.h>
+
+#include "loc_core.h"
+#include "loc_utils.h"
+
+#include "rest/rest_services_wifi.h"
+
+LOG_MODULE_DECLARE(location, CONFIG_LOCATION_LOG_LEVEL);
+
+BUILD_ASSERT(
+	IS_ENABLED(CONFIG_LOCATION_METHOD_WIFI_SERVICE_NRF_CLOUD) ||
+	IS_ENABLED(CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE) ||
+	IS_ENABLED(CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK),
+	"At least one WiFi positioning service must be enabled");
+
+extern location_event_handler_t event_handler;
+extern struct loc_event_data current_event_data;
+
+struct k_work_args {
+	struct k_work work_item;
+	struct loc_wifi_config wifi_config;
+	int64_t starting_uptime_ms;
+
+};
+
+static struct net_if *wifi_iface;
+
+static struct k_work_args method_wifi_start_work;
+
+static bool running;
+
+static uint32_t current_scan_result_count;
+static uint32_t latest_scan_result_count;
+
+struct method_wifi_scan_result {
+	char mac_addr_str[WIFI_MAC_ADDR_STR_LEN + 1];
+	char ssid_str[WIFI_SSID_MAX_LEN + 1];
+	uint8_t channel;
+	int8_t rssi;
+};
+
+static struct method_wifi_scan_result
+	latest_scan_results[CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT];
+static K_SEM_DEFINE(wifi_scanning_ready, 0, 1);
+
+/******************************************************************************/
+
+static int method_wifi_scanning_start(void)
+{
+	int ret;
+
+	LOG_DBG("Triggering start of WiFi scanning");
+
+	latest_scan_result_count = 0;
+	current_scan_result_count = 0;
+
+	if (wifi_iface != NULL) {
+		ret = net_mgmt(NET_REQUEST_WIFI_SCAN, wifi_iface, NULL, 0);
+		if (ret) {
+			LOG_ERR("Failed to initiate WiFi scanning: %d", ret);
+			ret = -1;
+		}
+	} else {
+		LOG_ERR("Failed to initiate WiFi scanning: no wifi device");
+		ret = -1;
+	}
+	return ret;
+}
+
+/******************************************************************************/
+
+static void method_wifi_handle_wifi_scan_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_scan_result *entry = (const struct wifi_scan_result *)cb->info;
+	struct method_wifi_scan_result *current;
+
+	current_scan_result_count++;
+	current = &latest_scan_results[current_scan_result_count - 1];
```

Comment:
no.. I don't see any wrong here.

Reply:
When the list becomes full, `current_scan_result_count` will be greater than `CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT`. In that case `current` points outside the array. The pointer isn't used in that case, but it would be better to move setting the `current` pointer inside the `if`, after the count has been checked.

Reply:
OK, now I see it. Thanks for the clarifications. It haven't caused problems, I have been facing the max many times.
But yeah, better move inside `if`